File: 111/main.c
#include <stdio.h>
#include <memory.h>

#define MAX(a,b) ((a) > (b) ? (a) : (b))

int main(int argc, char* argv[])
{
   int i, j, n, order;
   int table[21][21];
   int events[2][20];

   memset(table, 0, sizeof(table));

   scanf("%d", &n);
   for (i = 1; i <= n; i++) {
      scanf("%d", &order);
      events[0][order-1] = i;
   }

   while (scanf("%d", &order) == 1) {
      events[1][order-1] = 1;
      for (i = 2; i <=n ; i++) {
         scanf("%d", &order);
         events[1][order-1] = i;
      }

      for (i = 0; i < n; i++) {
         for (j = 0; j < n; j++) {
            if (events[0][i] == events[1][j]) {
               table[i+1][j+1] = table[i][j] + 1;
            } else {
               table[i+1][j+1] = MAX(table[i][j+1], table[i+1][j]);
            }
         }
      }
      printf("%d\n", table[i][j]);
   }

   return 0;
}

File: 111/Main.java
import java.io.*;
import java.util.*;

public class Main {
   public static void main(String args[]) throws IOException {
      int n, order;
      int[][] table;
      int[][] events;
      String line;
      StringTokenizer tokenizer;
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

      n = Integer.parseInt(in.readLine());
      table = new int[n+1][n+1];
      events = new int[2][n];

      line = in.readLine();
      readEvents(events[0], line);

      while ((line = in.readLine()) != null) {
         readEvents(events[1], line);

         int score = 0;
         for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
               if (events[0][i] == events[1][j]) {
                  table[i+1][j+1] = table[i][j] + 1;
               } else {
                  table[i+1][j+1] = Math.max(table[i][j+1], table[i+1][j]);
               }
               score = table[i+1][j+1];
            }
         }
         System.out.println(score);
      }
   }

   private static void readEvents(int[] events, String sequence) {
      int order;
      int index = 1;
      StringTokenizer tokenizer = new StringTokenizer(sequence);

      while (tokenizer.hasMoreTokens()) {
         order = Integer.parseInt(tokenizer.nextToken());
         events[order-1] = index;
         ++index;
      }
   }
}
